src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs

```diff
@@ -170,109 +200,118 @@ private SymbolicRegexMatcher(SymbolicRegexNode<TSet> rootNode, int captureCount,
+                reverseInitialStates[charKind] = GetOrCreateStateUnsafe(_reversePattern, charKind);
             }
             _reverseInitialStates = reverseInitialStates;
 
-            // Initialize our fast-lookup for determining the character kind of ASCII characters.
-            // This is only required when the pattern contains anchors, as otherwise there's only
-            // ever a single kind used.
-            if (_pattern._info.ContainsSomeAnchor)
+            // Maps a minterm ID to a character kind
+            uint CalculateMintermIdKind(int mintermId)
             {
-                var asciiCharKinds = new uint[128];
-                for (int i = 0; i < asciiCharKinds.Length; i++)
+                if (!_pattern._info.ContainsSomeAnchor)
                 {
-                    TSet set;
-                    uint charKind;
+                    return CharKind.General;
+                }
 
-                    if (i == '\n')
-                    {
-                        set = _builder._newLineSet;
-                        charKind = CharKind.Newline;
-                    }
-                    else
-                    {
-                        set = _builder._wordLetterForBoundariesSet;
-                        charKind = CharKind.WordLetter;
-                    }
+                // A minterm ID of -1 represents the positions before the first and after the last character
+                // in the input.
+                if (mintermId == -1)
+                {
+                    return CharKind.BeginningEnd;
+                }
+
+                // A minterm ID of minterms.Length represents a \n at the very end of input, which is matched
+                // by the \Z anchor.
+                if ((uint)mintermId == (uint)_minterms.Length)
+                {
+                    return CharKind.NewLineS;
+                }
+
+                TSet minterm = _minterms[mintermId];
 
-                    asciiCharKinds[i] = _builder._solver.And(GetMinterm(i), set).Equals(_builder._solver.Empty) ? 0 : charKind;
+                // Examine the minterm to figure out its character kind
+                if (_builder._newLineSet.Equals(minterm))
+                {
+                    // The minterm is a new line character
+                    return CharKind.Newline;
+                }
+                else if (!Solver.IsEmpty(Solver.And(_builder._wordLetterForBoundariesSet, minterm)))
+                {
+                    Debug.Assert(Solver.IsEmpty(Solver.And(Solver.Not(_builder._wordLetterForBoundariesSet), minterm)));
+                    // The minterm is a subset of word letters as considered by \b and \B
+                    return CharKind.WordLetter;
+                }
+                else
+                {
+                    // All other minterms belong to the general kind
+                    return CharKind.General;
```

Comment:
Maybe put the majority of the method into a `if (_pattern._info.ContainsSomeAnchor)` block so that we can dedup the `return CharKind.General;` to exist just once at the end?